components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java

```diff
@@ -3154,6 +3163,57 @@ public void addDocumentationContent(API api, String documentationName, String te
         }
     }
 
+    /**
+     * Updates a visibility of the documentation
+     *
+     * @param api               API
+     * @param documentation    Documentation
+     * @throws APIManagementException if failed to update visibility
+     */
+    private void updateDocVisibility(API api, Documentation documentation) throws APIManagementException {
+        try {
+            GenericArtifactManager artifactManager = APIUtil.getArtifactManager(registry,APIConstants.DOCUMENTATION_KEY);
+            if (artifactManager == null) {
+                String errorMessage = "Artifact manager is null when updating documentation of API " +
+                        api.getId().getApiName();
+                throw new APIManagementException(errorMessage);
+            }
+
+            GenericArtifact artifact = artifactManager.getGenericArtifact(documentation.getId());
+            String[] authorizedRoles = new String[0];
+            String visibleRolesList = api.getVisibleRoles();
+            if (visibleRolesList != null) {
+                authorizedRoles = visibleRolesList.split(",");
+            }
+
+            int tenantId;
+            String tenantDomain =
+                    MultitenantUtils.getTenantDomain(APIUtil.replaceEmailDomainBack(api.getId().getProviderName()));
+            try {
+                tenantId = getTenantId(tenantDomain);
+
+                GenericArtifact updateApiArtifact = APIUtil.createDocArtifactContent(artifact, api.getId(), documentation);
+                artifactManager.updateGenericArtifact(updateApiArtifact);
+                APIUtil.clearResourcePermissions(artifact.getPath(), api.getId(), tenantId);
+
+                APIUtil.setResourcePermissions(api.getId().getProviderName(), api.getVisibility(), authorizedRoles,
+                        artifact.getPath(), registry);
+
+                String docFilePath = artifact.getAttribute(APIConstants.DOC_FILE_PATH);
+                if (org.apache.commons.lang.StringUtils.isEmpty(docFilePath)) {
+                    int startIndex = docFilePath.indexOf("governance") + "governance".length();
+                    String filePath = docFilePath.substring(startIndex, docFilePath.length());
+                    APIUtil.setResourcePermissions(api.getId().getProviderName(), api.getVisibility(),
+                            authorizedRoles, filePath, registry);
+                }
+            } catch (UserStoreException e) {
+                throw new APIManagementException("Error in retrieving Tenant Information while adding api :"
+                        + api.getId().getApiName(), e);
```

Comment:
This operation is not related to API creation right? This is related to doc visibility update. Please modify the error messages accrdingly.

Reply:
Fixed with ec30343